// hexo-component-inferno/lib/view/plugin/gallery.js
"use strict";

var _inferno = require("inferno");

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var _require = require('inferno'),
    Component = _require.Component,
    Fragment = _require.Fragment;

var _require2 = require('../../util/cache'),
    cacheComponent = _require2.cacheComponent;

var Gallery =
/*#__PURE__*/
function (_Component) {
  _inherits(Gallery, _Component);

  function Gallery() {
    _classCallCheck(this, Gallery);

    return _possibleConstructorReturn(this, _getPrototypeOf(Gallery).apply(this, arguments));
  }

  _createClass(Gallery, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          head = _this$props.head,
          lightGallery = _this$props.lightGallery,
          justifiedGallery = _this$props.justifiedGallery;

      if (head) {
        return (0, _inferno.createFragment)([(0, _inferno.createVNode)(1, "link", null, null, 1, {
          "rel": "stylesheet",
          "href": lightGallery.cssUrl
        }), (0, _inferno.createVNode)(1, "link", null, null, 1, {
          "rel": "stylesheet",
          "href": justifiedGallery.cssUrl
        })], 4);
      }

      var js = "document.addEventListener('DOMContentLoaded', () => {\n            if (typeof $.fn.lightGallery === 'function') {\n                $('.article').lightGallery({ selector: '.gallery-item' });\n            }\n            if (typeof $.fn.justifiedGallery === 'function') {\n                if ($('.justified-gallery > p > .gallery-item').length) {\n                    $('.justified-gallery > p > .gallery-item').unwrap();\n                }\n                $('.justified-gallery').justifiedGallery();\n            }\n        });";
      return (0, _inferno.createFragment)([(0, _inferno.createVNode)(1, "script", null, null, 1, {
        "src": lightGallery.jsUrl,
        "defer": true
      }), (0, _inferno.createVNode)(1, "script", null, null, 1, {
        "src": justifiedGallery.jsUrl,
        "defer": true
      }), (0, _inferno.createVNode)(1, "script", null, null, 1, {
        "dangerouslySetInnerHTML": {
          __html: js
        }
      })], 4);
    }
  }]);

  return Gallery;
}(Component);

module.exports = cacheComponent(Gallery, 'plugin.gallery', function (props) {
  var head = props.head,
      helper = props.helper;
  return {
    head: head,
    lightGallery: {
      jsUrl: helper.cdn('lightgallery', '1.6.8', 'dist/js/lightgallery.min.js'),
      cssUrl: helper.cdn('lightgallery', '1.6.8', 'dist/css/lightgallery.min.css')
    },
    justifiedGallery: {
      jsUrl: helper.cdn('justifiedGallery', '3.7.0', 'dist/js/jquery.justifiedGallery.min.js'),
      cssUrl: helper.cdn('justifiedGallery', '3.7.0', 'dist/css/justifiedGallery.min.css')
    }
  };
});
