// hexo-component-inferno/lib/util/yaml.js
"use strict";

var yaml = require('js-yaml');

var YamlType = require('js-yaml/lib/js-yaml/type');

var YamlSchema = require('js-yaml/lib/js-yaml/schema'); // output null as empty in yaml


var YAML_SCHEMA = new YamlSchema({
  include: [require('js-yaml/lib/js-yaml/schema/default_full')],
  implicit: [new YamlType('tag:yaml.org,2002:null', {
    kind: 'scalar',
    resolve: function resolve(data) {
      if (data === null) {
        return true;
      }

      var max = data.length;
      return max === 1 && data === '~' || max === 4 && (data === 'null' || data === 'Null' || data === 'NULL');
    },
    construct: function construct() {
      return null;
    },
    predicate: function predicate(object) {
      return object === null;
    },
    represent: {
      empty: function empty() {
        return '';
      }
    },
    defaultStyle: 'empty'
  })]
});
module.exports = {
  parse: function parse(str) {
    return yaml.safeLoad(str);
  },
  stringify: function stringify(object) {
    return yaml.safeDump(object, {
      indent: 4,
      lineWidth: 1024,
      schema: YAML_SCHEMA
    });
  }
};
